File: app/src/main/java/com/techventus/wikipedianews/logging/Logger.java
package com.techventus.wikipedianews.logging;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.util.Set;

/**
 * Created by josephmalone on 6/15/15.
 */
public class Logger
{
	// Prefix to help filtering of logcat for output from our app
	private static final String PREFIX = "#WIKI# ";
	private static final String AT = "@";
	private static final String PARENTHETIC_NULL = "(null)";
	// ***Don't change this*** it's now set on app startup based on the debuggable flag in the manifest
	private static boolean showlog = false;
	private static boolean showErrorLog = false;

	/**
	 * Return whether logging is enabled (implies the debuggable flag is set)
	 *
	 * @return true if logging is enabled
	 */
	public static Boolean getShowlog()
	{
		return showlog;
	}

	public static void setEnabled(Boolean enabled)
	{
		showlog = enabled;
		i("Logger", "Logger ENABLED.");
	}

	public static void setShowErrorEnabled(boolean enabled){
		showErrorLog = enabled;
	}

	public static void e(String TAG, String message, Throwable e)
	{
		if (showlog || showErrorLog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.e(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message), e);
		}
	}

	public static void e(String TAG, String message)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.e(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message));
		}
	}

	public static void v(String TAG, String message)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.v(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message));
		}
	}

	public static void w(String TAG, String message)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.w(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message));
		}
	}

	public static void i(String TAG, String message)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.i(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message));
		}
	}

	public static void d(String TAG, String message)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			Log.d(TAG, PREFIX + ((message == null) ? PARENTHETIC_NULL : message));
		}
	}

	public static void logExtras(String TAG, Context context, Intent src)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			if (src == null || src.getExtras() == null)
			{
				return;
			}
			Set<String> keys = src.getExtras().keySet();
			for (String key : keys)
			{
				d(TAG, String.format(context + ", intent extra %s=%s", key, src.getExtras().get(key).toString()));
			}
		}
	}


	/**
	 * Logs a stack trace to TTY.
	 *
	 * @param TAG              Debug tag to display in log output
	 * @param e                Throwable to display
	 */
	public static void logStackTrace(String TAG, Throwable e)
	{
		if (showlog)
		{
			TAG += AT + Thread.currentThread().getName();
			// Legend tells of NULL exceptions roaming the world of Java...
			if (e != null)
			{
				// Display exception type.
				d(TAG, e.toString());

				// Display stack trace lines.
				for (StackTraceElement elem : e.getStackTrace())
				{
					// Copies the format of Exception.printStackTrace().
					d(TAG, "at " + elem.toString());
				}
			}
			else
			{
				d(TAG, "NULL exception passed to logStackTrace(). This should never happen!");
			}
		}
	}

}

File: app/src/main/java/com/techventus/wikipedianews/logging/Logger.kt
package com.techventus.wikipedianews.logging

import android.content.Context
import android.content.Intent
import android.util.Log

/**
 * Created by josephmalone on 6/15/15.
 */
object Logger {
    // Prefix to help filtering of logcat for output from our app
    private const val PREFIX = "#WIKI# "
    private const val AT = "@"
    private const val PARENTHETIC_NULL = "(null)"

    // ***Don't change this*** it's now set on app startup based on the debuggable flag in the manifest
    @JvmField
    var showlog = false

    @JvmField
    var showErrorLog = false

    /**
     * Return whether logging is enabled (implies the debuggable flag is set)
     *
     * @return true if logging is enabled
     */
    @JvmStatic
    fun getShowlog(): Boolean {
        return showlog
    }

    @JvmStatic
    fun setEnabled(enabled: Boolean) {
        showlog = enabled
        i("Logger", "Logger ENABLED.")
    }

    @JvmStatic
    fun setShowErrorEnabled(enabled: Boolean) {
        showErrorLog = enabled
    }

    @JvmStatic
    fun e(tag: String, message: String?, e: Throwable?) {
        var tag = tag
        if (showlog || showErrorLog) {
            tag += AT + Thread.currentThread().name
            Log.e(tag, PREFIX + (message ?: PARENTHETIC_NULL), e)
        }
    }

    @JvmStatic
    fun e(tag: String, message: String?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            Log.e(tag, PREFIX + (message ?: PARENTHETIC_NULL))
        }
    }

    @JvmStatic
    fun v(tag: String, message: String?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            Log.v(tag, PREFIX + (message ?: PARENTHETIC_NULL))
        }
    }

    @JvmStatic
    fun w(tag: String, message: String?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            Log.w(tag, PREFIX + (message ?: PARENTHETIC_NULL))
        }
    }

    @JvmStatic
    fun i(tag: String, message: String?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            Log.i(tag, PREFIX + (message ?: PARENTHETIC_NULL))
        }
    }

    @JvmStatic
    fun d(tag: String, message: String?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            Log.d(tag, PREFIX + (message ?: PARENTHETIC_NULL))
        }
    }

    @JvmStatic
    fun logExtras(tag: String, context: Context, src: Intent?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            if (src?.extras == null) {
                return
            }
            val keys = src.extras!!.keySet()
            for (key in keys) {
                d(tag, String.format("$context, intent extra %s=%s", key, src.extras!![key].toString()))
            }
        }
    }

    /**
     * Logs a stack trace to TTY.
     *
     * @param tag              Debug tag to display in log output
     * @param e                Throwable to display
     */
    @JvmStatic
    fun logStackTrace(tag: String, e: Throwable?) {
        var tag = tag
        if (showlog) {
            tag += AT + Thread.currentThread().name
            // Legend tells of NULL exceptions roaming the world of Java...
            if (e != null) {
                // Display exception type.
                d(tag, e.toString())

                // Display stack trace lines.
                for (elem in e.stackTrace) {
                    // Copies the format of Exception.printStackTrace().
                    d(tag, "at $elem")
                }
            } else {
                d(tag, "NULL exception passed to logStackTrace(). This should never happen!")
            }
        }
    }
}